### PR TITLE
Remove unused type checks 

### LINKED ISSUE #86: Optimize away unused types for `console.log`/etc

### PR DESCRIPTION
**Pull request by BobVarioa**
Fixes #86. 

Before:
```sh
$ ./porf -e "console.log(1)" -b --no-run
wasm size: 46019 bytes
```

After:
```sh
$ ./porf -e "console.log(1)" -b --no-run
wasm size: 37735 bytes
```

For comparison: 
```sh
$ ./porf -e "Porffor.miniLog(1)" -b --no-run
wasm size: 23539 bytes
$ ./porf -e "Porffor.numberLog(1)" -b --no-run
wasm size: 115 bytes
```

So by no means magic, but a roughly 20% size reduction for a very common method. Likely could be optimized further via #64 and #119. 

This PR also adds `__Porffor_printArray` to reduce some of the code duplication in `console.log`

### CODE REVIEW

File: compiler/builtins/array.ts [hunk 1]
```diff
@@ -570,16 +570,14 @@ export const __Array_prototype_sort = (_this: any[], callbackFn: any) => {
 
       // 23.1.3.30.2 CompareArrayElements (x, y, comparefn)
       // https://tc39.es/ecma262/#sec-comparearrayelements
-      const xt: i32 = Porffor.rawType(x);
```
Comment (CanadaHonk) [inline compiler/builtins/array.ts:573]: unrelated changes?

File: compiler/opt.js [hunk 2]
```diff
@@ -4,6 +4,39 @@ import { read_signedLEB128, read_ieee754_binary64 } from './encoding.js';
...
+    case 'Uint8Array':
+    case 'Int8Array':
+    case 'Uint8ClampedArray':
+    case 'Uint16Array':
+    case 'Int16Array':
+    case 'Uint32Array':
+    case 'Int32Array':
+    case 'Float32Array':
+    case 'Float64Array':
+      return funcs.some(x => x.name === type);
```
Comment (CanadaHonk) [inline compiler/opt.js:33]: This would break eg `new Uint8Array(8).buffer`?

File: compiler/codegen.js [hunk 3]
```diff
@@ -3533,6 +3585,34 @@ const generateUpdate = (scope, decl, _global, _name, valueUnused = false) => {
 const generateIf = (scope, decl) => {
   const out = truthy(scope, generate(scope, decl.test), getNodeType(scope, decl.test), false, true);
 
+  // todo: support ||, not just fastOr
```
Comment (CanadaHonk) [inline compiler/codegen.js:3588]: ngl I don't really like all this added complexity just for some built-in cases in `if`s

File: compiler/codegen.js [hunk 4]
```diff
@@ -2671,36 +2671,86 @@ const brTable = (input, bc, returns) => {
...
+  if (!Array.isArray(bcArr)) {
+    bcArr = Object.entries(bc);
+  } else {
+    bc = Object.fromEntries(bcArr);
+  }
 
-  const tmp = localTmp(scope, '#typeswitch_tmp' + (Prefs.typeswitchUniqueTmp ? uniqId() : ''), Valtype.i32);
+  typeswitchDepth++;
+
+  const tmp = localTmp(scope, `#typeswitch_tmp${typeswitchDepth}`, Valtype.i32);
```
Comment (CanadaHonk) [inline compiler/codegen.js:2702]: Don't break the pref?

File: compiler/codegen.js [hunk 5]
```diff
@@ -2671,36 +2671,86 @@ const brTable = (input, bc, returns) => {
...
-    // if type == x
-    out.push([ Opcodes.local_get, tmp ]);
-    out.push(...number(x, Valtype.i32));
-    out.push([ Opcodes.i32_eq ]);
-
-    out.push([ Opcodes.if, Blocktype.void, `TYPESWITCH|${TYPE_NAMES[x]}` ]);
-    out.push(...bc[x]);
-    out.push([ Opcodes.br, 1 ]);
-    out.push([ Opcodes.end ]);
+    if (allowFallThrough) {
```
Comment (CanadaHonk) [inline compiler/codegen.js:2713]: add comments explaining this please?

**Review by CanadaHonk — CHANGES_REQUESTED** [review]

File: compiler/codegen.js (hunk 3 above)
Comment (BobVarioa) [reply compiler/codegen.js:3588]: Should I just optimize ifs that have one type check, or remove this completely?

File: compiler/opt.js (hunk 2 above)
Comment (BobVarioa) [reply compiler/opt.js:33]: I'm not sure what you mean? That'll still call the constructor, plus this is what the code was doing before, just separated into another function

Comment (BobVarioa) [reply compiler/opt.js:33]: Ignore me, I didn't see that part of the code, I'll remove arraybuffer and sharedarray buffer from the list

File: compiler/codegen.js (hunk 3 above)
Comment (CanadaHonk) [reply compiler/codegen.js:3588]: I think for now this can just be removed

Comment (BobVarioa) [reply compiler/codegen.js]: Done!

File: compiler/codegen.js (hunk 4 above)
Comment (BobVarioa) [reply compiler/codegen.js]: Woops 😓

File: compiler/codegen.js (hunk 5 above)
Comment (BobVarioa) [reply compiler/codegen.js:2715]: Done!

File: compiler/codegen.js (hunk 3 above)
Comment (CanadaHonk) [reply compiler/codegen.js]: A lot of builtins are still unneededly changed for this?

Comment (BobVarioa) [reply compiler/codegen.js]: I will say especially once #116 is merged those are really moreso optimizations, so we don't introduce an unnecessary local. But, I'll revert those changes regardless

Comment (CanadaHonk) [reply compiler/codegen.js]: I think it is nicer code generally fwiw, we should have more general optimizations later which handles this sort of stuff rather than focusing on a few select scenarios imo

File: compiler/builtins/console.ts [hunk 6]
```diff
@@ -89,8 +89,29 @@ export const __Porffor_miniLog = (arg: any) => {
 export const __Porffor_print = (arg: any, colors: boolean = true) => {
   // todo: Symbol.toStringTag could reduce duplication here
 
-  const t: i32 = Porffor.rawType(arg);
-  switch (t) {
+  // note: this doesn't have access to the upper scope!! do not use any variables from up there
+  const __Porffor_printArray = (arg: any, colors: boolean, length: boolean = false) => {
```
Comment (CanadaHonk) [inline compiler/builtins/console.ts:93]: Could you put this at the top-level and exported like the others please?

File: compiler/builtins/console.ts [hunk 7]
```diff
@@ -89,8 +89,29 @@ export const __Porffor_miniLog = (arg: any) => {
...
+      printStatic('[]');
+    } else {
+      printStatic('[ ');
+      for (let i: i32 = 0; i <= arrLen; i++) {
+        __Porffor_print(arg[i], colors);
+        if (i != arrLen) printStatic(', ');
+      }
+      printStatic(' ]');
+    }
+  }
```
Comment (CanadaHonk) [inline compiler/builtins/console.ts:112]: style: `;`

File: compiler/builtins/console.ts [hunk 8]
```diff
@@ -89,8 +89,29 @@ export const __Porffor_miniLog = (arg: any) => {
 export const __Porffor_print = (arg: any, colors: boolean = true) => {
   // todo: Symbol.toStringTag could reduce duplication here
 
-  const t: i32 = Porffor.rawType(arg);
```
Comment (CanadaHonk) [inline compiler/builtins/console.ts:92]: Could you also bring this back please? :sweat_smile: sorry

File: compiler/opt.js [hunk 9]
```diff
@@ -377,18 +409,6 @@ export default (funcs, globals, pages, tags, exceptions) => {
           continue;
         }
 
-        if (false && i === wasm.length - 1 && inst[0] === Opcodes.return) {
```
Comment (CanadaHonk) [inline compiler/opt.js:380]: Why remove this?

File: compiler/codegen.js [hunk 10]
```diff
@@ -4137,6 +4197,42 @@ const generateSwitch = (scope, decl) => {
 
   depth.push('switch');
 
+  if (
```
Comment (CanadaHonk) [inline compiler/codegen.js:4200]: I thought we agreed to remove all this for now/this PR?

**Review by CanadaHonk — CHANGES_REQUESTED** [review]

File: compiler/builtins/console.ts (hunk 6 above)
Comment (BobVarioa) [reply compiler/builtins/console.ts:93]: The circular dependency breaks it if I do that unfortunately

File: compiler/opt.js (hunk 9 above)
Comment (BobVarioa) [reply compiler/opt.js:380]: Woops, I need to find the VSCode extension that's doing that, it removed "dead code" on save.

File: compiler/codegen.js (hunk 10 above)
Comment (BobVarioa) [reply compiler/codegen.js:4200]: Did you mean the stuff for `generateSwitch` too? That's the meat of this PR in all honesty

File: compiler/builtins/console.ts (hunk 6 above)
Comment (CanadaHonk) [reply compiler/builtins/console.ts:93]: oh oops, seems fine then :+1:

File: compiler/codegen.js (hunk 10 above)
Comment (CanadaHonk) [reply compiler/codegen.js:4200]: I may have missed where this was oops, nvm!

File: compiler/opt.js (hunk 9 above)
Comment (CanadaHonk) [reply compiler/opt.js:380]: Ah np, just wondered if it was intentional heh

**Review by CanadaHonk — APPROVED** [review]
Looks good, thanks!

**Comment by CanadaHonk** [comment]
Could you fix the merge conflicts? Sorry :sweat:

**Comment by BobVarioa** [comment]
> Could you fix the merge conflicts? Sorry 😓

image
I might be wrong but it doesn't have any?

**Comment by CanadaHonk** [comment]
Oh nvm I can merge but not rebase :saluting_face:
